File: typescript/artifactLoader.ts
type RequireLike = ((id: string) => unknown) & { main?: unknown };
type ModuleLike = object;
type ProcessLike = { cwd: () => string };

declare const require: RequireLike;
declare const module: ModuleLike;
declare const process: ProcessLike;

export type ArtifactCatalog = {
  soul: string[];
  skills: string[];
};

type DirEntryLike = {
  name: string;
  isDirectory: () => boolean;
};

const fs = require("fs") as {
  readdirSync: (path: string, options: { withFileTypes: boolean }) => DirEntryLike[];
  statSync: (path: string) => { isFile: () => boolean };
};
const path = require("path") as { join: (...parts: string[]) => string };

function collect(parentDir: string, markerFile: string, repoRoot: string): string[] {
  let entries: DirEntryLike[];
  try {
    entries = fs.readdirSync(parentDir, { withFileTypes: true });
  } catch {
    return [];
  }

  return entries
    .filter((entry) => entry.isDirectory())
    .map((entry) => path.join(parentDir, entry.name))
    .filter((dirPath: string) => {
      const marker = path.join(dirPath, markerFile);
      try {
        return fs.statSync(marker).isFile();
      } catch {
        return false;
      }
    })
    .map((dirPath: string) => path.join(dirPath, markerFile).slice(repoRoot.length).replace(/^[/\\]/, "").replace(/\\/g, "/"));
}

export function discoverArtifacts(repoRoot: string): ArtifactCatalog {
  return {
    soul: collect(path.join(repoRoot, "soul"), "SOUL.md", repoRoot).sort(),
    skills: collect(path.join(repoRoot, ".agents", "skills"), "SKILL.md", repoRoot).sort(),
  };
}

if (require.main === module) {
  const catalog = discoverArtifacts(process.cwd());
  console.log("soul:");
  for (const s of catalog.soul) console.log(`- ${s}`);
  console.log("skills:");
  for (const skill of catalog.skills) console.log(`- ${skill}`);
}

File: python/artifact_loader.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class ArtifactCatalog:
    soul: list[str]
    skills: list[str]


def discover_artifacts(repo_root: Path) -> ArtifactCatalog:
    soul_root = repo_root / "soul"
    skills_root = repo_root / ".agents" / "skills"

    soul = sorted(
        str(path.parent.relative_to(repo_root))
        for path in soul_root.glob("*/SOUL.md")
        if path.is_file()
    )

    skills = sorted(
        str(path.parent.relative_to(repo_root))
        for path in skills_root.glob("*/SKILL.md")
        if path.is_file()
    )

    return ArtifactCatalog(soul=soul, skills=skills)


if __name__ == "__main__":
    # Script lives in <repo>/python, so parent[1] is repository root.
    catalog = discover_artifacts(Path(__file__).resolve().parents[1])
    print("soul:")
    for s in catalog.soul:
        print(f"- {s}")
    print("skills:")
    for skill in catalog.skills:
        print(f"- {skill}")
